rafayeljamalyandev/Interview-ToDo-App
# Candidate improvements

This pull request implements several key features and improvements for the Todos API, including authentication, authorization, and integration tests. Additionally, it provides a Postman collection for testing the endpoints. The codebase has been refactored to follow clean architecture and modular design principles, ensuring better maintainability and scalability.

Changes:
Authentication & Authorization:
Auth Module:
Created endpoints for user registration (POST /auth/register), login (POST /auth/login), and token refresh (POST /auth/refresh).
Integrated JWT authentication to secure endpoints.
Added AtGuard and RtGuard for access token and refresh token validation, respectively.
Custom Decorators:
Added @GetCurrentUser and @GetCurrentUserId decorators to easily access the current user's data within controller methods.
Added a @Public decorator to mark routes as public (i.e., not requiring authentication).

Todos Module:
Created endpoints to perform CRUD operations on the Todo list:
POST /todos: Create a new Todo.
GET /todos: List all Todos.
GET /todos/:userId: List Todos for a specific user.
PUT /todos/:id: Update a Todo by ID.
Integrated guards to protect the routes, ensuring that only authorized users can perform certain actions.
Added Prisma ORM integration for database management, including methods for connecting and disconnecting from the database.

Testing:
Integration Tests:
Created end-to-end tests for the Todos API, ensuring the routes work as expected.
Mocked the AtGuard to avoid JWT dependency during testing.
Postman Collection:
Added a Postman collection to allow manual and automated testing of the auth and Todos endpoints.
Includes requests for registration, login, refresh token, and CRUD operations on Todos.

Code Cleanup & Maintenance:
Defined a custom PrismaService to handle database connection lifecycle.
Enhanced type safety with custom Express request types to access user information.

Testing Instructions:

Import the provided Postman collection (.postman-collection.json) into Postman.
Set up environment variables for access_token and refresh_token based on the responses from the login and refresh endpoints.
Test the authentication flows (/auth/register, /auth/login, /auth/refresh) and Todos CRUD operations (/todos routes).

Notes:

Ensure the database is properly set up and the connection URL is correctly configured in the environment variables.
The changes include mock testing for JWT authentication, so real tokens are not required in the test suite.
